not-context-manager tests fail with astroid 2.0

### Steps to reproduce
1. Use astroid 2.0 as in the following commit https://github.com/PyCQA/pylint/commit/88d336ec29ad046f65c64f56f68637c321d744b2
2. Run the test suite using `tox`


### Current behavior

The tests for the `not-context-manager` check will fail with:
```
>           pytest.fail('\n'.join(msg))
E           Failed: Wrong results for file "not_context_manager":
E           
E           Unexpected in testdata:
E            121: not-context-manager
E            125: not-context-manager
```

### Expected behavior

For the tests to not fail. I *think* this is caused by https://github.com/PyCQA/astroid/commit/53ab62402b2842dbe79cfbd91f3d698d74200485, which modifies the way (in a correct way) how contexts works. We were actually asking for some breakage around here https://github.com/PyCQA/pylint/blob/72df0403c84b3ac6deb7cb4023912b58321c3181/pylint/checkers/typecheck.py#L1148, where we were relying on inspecting the context's path to see if we find the scope of the context manager. We should probably find another way to infer that, rather than inspecting something as brittle as the context inference path.

CC @brycepg

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -37,6 +37,7 @@ import operator
 import re
 import shlex
 import sys
+import types
 
 import astroid
 import astroid.context
@@ -77,6 +78,14 @@ def _unflatten(iterable):
         elif elem and not index:
             # We're interested only in the first element.
             yield elem
+
+def _flatten_container(iterable):
+    # Flatten nested containers into a single iterable
+    for item in iterable:
+        if isinstance(item, (list, tuple, types.GeneratorType)):
+            yield from _flatten_container(item)
+        else:
+            yield item
 
 
 def _is_owner_ignored(owner, name, ignored_classes, ignored_modules):
@@ -1145,8 +1154,12 @@ accessed. Python regular expressions are accepted.'}
                 # manager and give up, otherwise emit not-context-manager.
                 # See the test file for not_context_manager for a couple
                 # of self explaining tests.
-                for path in filter(None, _unflatten(context.path)):
-                    scope = path.scope()
+
+                # Retrieve node from all previusly visited nodes in the the inference history
+                context_path_names = filter(None, _unflatten(context.path))
+                inferred_paths = _flatten_container(path.infer() for path in context_path_names)
+                for inf_path in inferred_paths:
+                    scope = inf_path.scope()
                     if not isinstance(scope, astroid.FunctionDef):
                         continue
                     if decorated_with(scope,

